From 49963dd72fe4de43b8cf356d50d29b7782cd4318 Mon Sep 17 00:00:00 2001
From: spatel <spatel@maxlinear.com>
Date: Thu, 10 Oct 2024 23:18:26 +0530
Subject: [PATCH] WLANRTSYS-61018 [MXL] Puma8: ACS 320MHz

    - Implemented 320MHz support based on EHT Operation IE parsing
    - Fixed seg0/seg1 interpretation for 2.4GHz and 5GHz
    - Fixed secondary frequency calculation for all frequency bands
---
 hostapd/mxl_config_proprietary.c     | 18 ++++++++++++++++--
 hostapd/mxl_ctrl_iface_proprietary.c |  2 +-
 src/ap/mxl_acs_proprietary.c         | 26 +++++++++++++++++++++++++-
 src/ap/mxl_config.h                  | 15 ++++++++++++++-
 4 files changed, 56 insertions(+), 5 deletions(-)

diff --git a/hostapd/mxl_config_proprietary.c b/hostapd/mxl_config_proprietary.c
index b16bc84b8..0c93ee83a 100644
--- a/hostapd/mxl_config_proprietary.c
+++ b/hostapd/mxl_config_proprietary.c
@@ -382,6 +382,14 @@ int mxl_config_check_smart_acs_config(struct hostapd_config *conf, struct hostap
 		mxl_conf->acs_chanlist_all_chans = val;
 	} else if (os_strcmp(buf, "acs_scan_mode") == 0) {
 		mxl_conf->acs_scan_mode = atoi(pos);
+	} else if (os_strcmp(buf, "acs_eht_mode") == 0) {
+		int val = atoi(pos);
+		if(!IS_VALID_RANGE(val, ACS_EHT_MODE_MIN, ACS_EHT_MODE_MAX)) {
+			wpa_printf(MSG_ERROR, "Invalid EHT mode, "
+				"must be in range %i..%i", ACS_EHT_MODE_MIN, ACS_EHT_MODE_MAX);
+			return 1;
+		}
+		mxl_conf->acs_eht_mode = (enum acs_eht_modes)val;
 	} else if (os_strcmp(buf, "scan_timeout") == 0) {
 		int val = atoi(pos);
 		if (!IS_VALID_RANGE(val, SCAN_TIMEOUT_MIN, SCAN_TIMEOUT_MAX)) {
@@ -398,6 +406,11 @@ int mxl_config_check_smart_acs_config(struct hostapd_config *conf, struct hostap
 			return 0;
 		}
 		mxl_conf->acs_bgscan_interval = val;
+	} else if (os_strcmp(buf, "acs_exclude_opclass_ch_list") == 0) {
+		if (hostapd_parse_opclass_ch_list(&mxl_conf->acs_exclude_opclass_ch_list, pos, &mxl_conf->acs_exclude_opclass_ch_count)) {
+			wpa_printf(MSG_ERROR, "Line %d: Invalid acs_exclude_opclass_ch_list value", line);
+			return 0;
+		}
 	} else {
 		return 0;
 	}
@@ -410,8 +423,8 @@ int mxl_hostapd_acs_config_defaults(struct hostapd_config *conf)
 {
 #ifdef CONFIG_ACS
 	int i;
-	int grp_priorities_throughput[ACS_NUM_GRP_PRIORITIES] = { 3, 2, 1, 0 };
-	int grp_priorities_reach[ACS_NUM_GRP_PRIORITIES] = { 0, 1, 2, 3 };
+	int grp_priorities_throughput[ACS_NUM_GRP_PRIORITIES] = { 4, 3, 2, 1, 0 };
+	int grp_priorities_reach[ACS_NUM_GRP_PRIORITIES] = { 0, 1, 2, 3, 4 };
 	int acs_penalty_factors[ACS_NUM_PENALTY_FACTORS] = {ACS_PEN_DEFAULT, ACS_PEN_DEFAULT, ACS_PEN_DEFAULT, ACS_PEN_DEFAULT_WBSS};
 	struct hostapd_bss_config *bss = conf->bss[0];
 #endif /* CONFIG_ACS */
@@ -513,6 +526,7 @@ void mxl_hostapd_config_free_acs(struct hostapd_config *conf)
 	os_free(mxl_conf->acs_grp_priorities_throughput);
 	os_free(mxl_conf->acs_grp_priorities_reach);
 	os_free(mxl_conf->acs_6g_opt_ch_list);
+	os_free(mxl_conf->acs_exclude_opclass_ch_list);
 #endif /* CONFIG_ACS */
 
 }
diff --git a/hostapd/mxl_ctrl_iface_proprietary.c b/hostapd/mxl_ctrl_iface_proprietary.c
index a016c9386..ce1cbbcfe 100644
--- a/hostapd/mxl_ctrl_iface_proprietary.c
+++ b/hostapd/mxl_ctrl_iface_proprietary.c
@@ -68,7 +68,7 @@ static int mxl_hostapd_ctrl_iface_acs_report(struct hostapd_iface *iface,
 
 		if (candidates[i].filled_mask & CHDATA_TX_POWER) {
 			ret = sprintf_s(buf + len, buflen - len, " pow=%d",
-					candidates[i].tx_power);
+					candidates[i].tx_power[PRI_20]);
 			if (ret <= 0)
 				break;
 			len += ret;
diff --git a/src/ap/mxl_acs_proprietary.c b/src/ap/mxl_acs_proprietary.c
index 263bdb4ed..97cdf337d 100644
--- a/src/ap/mxl_acs_proprietary.c
+++ b/src/ap/mxl_acs_proprietary.c
@@ -1943,6 +1943,22 @@ static int acs_unii4_chan_disabled(struct hostapd_iface *iface, struct hostapd_c
 	return false;
 }
 
+static int is_excluded_opclass_chan(struct hostapd_iface *iface, int channel, int bandwidth)
+{
+	int i, op_band;
+	struct mxl_hostapd_config *mxl_conf = &iface->conf->mxl_conf;
+
+	for (i = 0; i < mxl_conf->acs_exclude_opclass_ch_count; i++) {
+		struct acs_exclude *op_list = &mxl_conf->acs_exclude_opclass_ch_list[i];
+
+		op_band = op_class_to_bandwidth(op_list->op_class);
+		if (op_band == bandwidth && op_list->channel == channel) {
+			return 1;
+		}
+	}
+	return 0;
+}
+
 static int acs_is_subchannel_allowed(struct hostapd_iface *iface, int channel)
 {
 	int j;
@@ -2073,6 +2089,11 @@ void add_channel_pair_candidates(struct hostapd_iface *iface, channel_pair *chan
 			    (acs_is_40_mhz_overlap_chan(&channel_pairs[i])))
 				continue;
 
+			if (is_excluded_opclass_chan(iface, channel_pairs[i].primary, 40) ||
+				is_excluded_opclass_chan(iface, channel_pairs[i].secondary, 40)) {
+				continue;
+			}
+
 			if (channel_pairs[i].primary < channel_pairs[i].secondary)
 				candidates[iface->mxl_iface.num_candidates].freq = chan->freq;
 			else
@@ -2239,7 +2260,8 @@ void mxl_acs_init_candidate_table(struct hostapd_iface *iface)
 
 		if (!acs_unii4_chan_disabled(iface, chan) &&
 		    (iface->conf->mxl_conf.acs_use24overlapped || (chan->chan == 1) ||
-		    (chan->chan == 6) || (chan->chan == 11) || (chan->freq >= 5000))) {
+		    (chan->chan == 6) || (chan->chan == 11) || (chan->freq >= 5000)) &&
+		    !is_excluded_opclass_chan(iface, chan->chan, 20)) {
 			if (chan->flag & HOSTAPD_CHAN_RADAR)
 				candidates[iface->mxl_iface.num_candidates].radar_affected = true;
 
@@ -4912,6 +4934,8 @@ int mxl_acs_check_bgscan(struct hostapd_iface *iface)
 	if (iface->state == HAPD_IFACE_ENABLED)
 		iface->mxl_iface.acs_reenable_iface = 1;
 
+	iface->conf->mxl_conf.acs_init_done = 0;
+
 	return mxl_acs_start(iface);
 }
 
diff --git a/src/ap/mxl_config.h b/src/ap/mxl_config.h
index b09039200..bd6769389 100644
--- a/src/ap/mxl_config.h
+++ b/src/ap/mxl_config.h
@@ -28,7 +28,7 @@ struct hostapd_bss_config;
 #define ACS_PEN_MAX                     1000
 #define ACS_PEN_DEFAULT                 100
 #define ACS_PEN_DEFAULT_WBSS            50
-#define ACS_NUM_GRP_PRIORITIES          4
+#define ACS_NUM_GRP_PRIORITIES          5
 #define ACS_MAX_CHANNELS                (233+1)
 #define ACS_SWITCH_THRESHOLD            100
 #define ACS_MIN_NUM_SCANS               1
@@ -43,6 +43,19 @@ typedef struct acs_chan {
         int width;
 } acs_chan;
 
+struct acs_exclude {
+	int op_class;
+	int channel;
+};
+
+enum acs_eht_modes {
+	ACS_EHT_MODE_MIN,
+	ACS_EHT_MODE_320_0 = ACS_EHT_MODE_MIN,
+	ACS_EHT_MODE_320_1,
+	ACS_EHT_MODE_320_2,
+	ACS_EHT_MODE_MAX = ACS_EHT_MODE_320_2
+};
+
 int mxl_config_check_smart_acs_config(struct hostapd_config *conf, struct hostapd_bss_config *bss, const char *buf, char* pos, int line);
 int mxl_hostapd_acs_config_defaults(struct hostapd_config *conf);
 void mxl_hostapd_config_fill_acs(struct hostapd_config *conf, char *pos);
-- 
2.43.0

